### PR TITLE
Refactor schedules - part 1

### PR DESCRIPTION
**Pull request by syoung-smallwisdom**
This includes both some older changes to support associating an upload with either an assessment **or** a session window.

Additionally, I started migrating the logic for randomized start times into the library. I have not hooked this up or finished adding testing all the way through, but it does include tests and model objects for all the mutation.  Since it's a lot of code, I felt that I needed to break it up.

### CODE REVIEW

File: bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/models/UserAvailabilityWindow.kt [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+     * Calculate evenly spaced session times throughout the repeat window.
+     */
+    internal fun evenSpacedSessionTimes(repeatTimeWindow: RepeatTimeWindow) : CalculatedScheduleWindows
+        = evenSpacedSessionTimes(repeatTimeWindow.count, repeatTimeWindow.expiration)
+
+    private fun evenSpacedSessionTimes(sessionsPerDay: Int, windowDuration: DateTimePeriod) : CalculatedScheduleWindows {
+        val totalAvailabilityInMinutes = availabilityInMinutes()
+        val desiredSessionWindow = totalAvailabilityInMinutes / sessionsPerDay
+        val minimumSessionWindow = windowDuration.totalMinutes
+        // if the availability is less that the required availability then use the minimumSessionWindow
```
Comment (Erin-Mounts) [inline bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/models/UserAvailabilityWindow.kt]:
```suggestion
        // if the availability is less than the required availability then use the minimumSessionWindow
```

File: bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AppConfigRepo.kt [hunk 2]
```diff
@@ -12,6 +13,8 @@ import org.sagebionetworks.bridge.kmm.shared.cache.ResourceDatabaseHelper
 import org.sagebionetworks.bridge.kmm.shared.cache.ResourceResult
 import org.sagebionetworks.bridge.kmm.shared.cache.ResourceType
 import org.sagebionetworks.bridge.kmm.shared.models.AppConfig
+import org.sagebionetworks.bridge.kmm.shared.models.ScheduleConfig
+import org.sagebionetworks.bridge.kmm.shared.models.scheduleConfig
```
Comment (Erin-Mounts) [inline bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AppConfigRepo.kt]: Looks like you added this twice (assuming case doesn’t matter)?

File: bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/ParticipantRepo.kt [hunk 3]
```diff
@@ -130,25 +166,44 @@ class ParticipantRepo(httpClient: HttpClient,
...
-        var clientData: JsonElement? = null,
         /* The user's email. */
         var email: String? = null,
-        var phone: Phone? = null
+        var phone: Phone? = null,
+
+        /* Data used to set up custom schedule times. */
+        internal var userScheduleData: UserScheduleData?,
+
+        // wrap the client data to allow both the app to store app-specific properties on the
```
Comment (Erin-Mounts) [inline bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/ParticipantRepo.kt]:
```suggestion
        // Wrap the client data to allow both the app to store app-specific properties on the
```

**Review by Erin-Mounts — APPROVED** [review]
Couple of nits/questions, otherwise looks good to me

File: bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AppConfigRepo.kt [hunk 4]
```diff
@@ -12,6 +12,8 @@ import org.sagebionetworks.bridge.kmm.shared.cache.ResourceDatabaseHelper
...
```
Comment (syoung-smallwisdom) [inline bridge-client/src/commonMain/kotlin/org/sagebionetworks/bridge/kmm/shared/repo/AppConfigRepo.kt]:
```suggestion
```

**Pull request merged**